File: 2015/cpp/day06.cpp
#include <cassert>
#include <vector>
#include <string>
#include <sstream>
#include <utility>
#include <fstream>
#include <array>
#include <functional>
#include <algorithm>

using namespace std;

using OpFunc = function<int(int)>;

enum class Type {
    On = 0,
    Off = 1,
    Toggle = 2,
};

struct Operation {
    Type type;
    pair<int, int> start;
    pair<int, int> end;

    explicit Operation(Type type, pair<int, int> start, pair<int, int> end)
        : type(type), start(std::move(start)), end(std::move(end)) {
    }
};

auto ParseNumbers(const string &s) -> pair<int, int> {
    stringstream ss(s);
    string tmp;
    vector<int> v;

    while (getline(ss, tmp, ',')) {
        v.push_back(stoi(tmp));
    }

    assert(v.size() == 2);

    return {v[0], v[1]};
}

auto Parse(const vector<string> &input) -> vector<Operation> {
    vector<Operation> ret;
    for (const auto &line : input) {
        stringstream ss(line);
        string tmp;
        vector<string> v;

        while (getline(ss, tmp, ' ')) {
            v.push_back(tmp);
        }

        auto type = v[1] == "on" ? Type::On : v[1] == "off" ? Type::Off : Type::Toggle;
        if (type == Type::Toggle) {
            ret.emplace_back(type, ParseNumbers(v[1]), ParseNumbers(v[3]));
        } else {
            ret.emplace_back(type, ParseNumbers(v[2]), ParseNumbers(v[4]));
        }
    }
    return ret;
}

auto CountLight(const vector<Operation> &input, const array<OpFunc, 3> &ops) -> int {
    vector<vector<int>> grid(1000, vector<int>(1000, false));

    for (const auto &op : input) {
        auto &[row1, col1] = op.start;
        auto &[row2, col2] = op.end;
        auto func = ops[static_cast<int>(op.type)];
        for (int i = row1; i <= row2; ++i) {
            for (int j = col1; j <= col2; ++j) {
                grid[i][j] = func(grid[i][j]);
            }
        }
    }

    int ret = 0;
    for (int i = 0; i < 1000; ++i) {
        for (int j = 0; j < 1000; ++j) {
            ret += grid[i][j];
        }
    }
    return ret;
}

auto Problem1(const vector<Operation> &input) -> int {
    const std::array<OpFunc, 3> ops = {
        [](int v) -> int { return 1; },
        [](int v) -> int { return 0; },
        [](int v) -> int { return v == 1 ? 0 : 1; },
    };

    return CountLight(input, ops);
}

auto Problem2(const vector<Operation> &input) -> int {
    const std::array<OpFunc, 3> ops = {
        [](int v) -> int { return v + 1; },
        [](int v) -> int { return max(v - 1, 0); },
        [](int v) -> int { return v + 2; },
    };

    return CountLight(input, ops);
}

void Test() {
    vector<Operation> input1 = {Operation(Type::On, {0, 0}, {999, 999})};
    assert(Problem1(input1) == 1'000'000);

    vector<Operation> input2 = {Operation(Type::Toggle, {0, 0}, {999, 0})};
    assert(Problem1(input2) == 1'000);

    vector<Operation> input3 = {Operation(Type::On, {0, 0}, {0, 0})};
    assert(Problem2(input3) == 1);
    vector<Operation> input4 = {Operation(Type::Toggle, {0, 0}, {999, 999})};
    assert(Problem2(input4) == 2'000'000);
}

int main() {
    Test();

    ifstream f("../input/day06.txt");
    vector<string> v;
    string tmp;
    while (getline(f, tmp)) {
        v.push_back(tmp);
    }

    auto input = Parse(v);
    int ret1 = Problem1(input);
    assert(ret1 == 400410);

    int ret2 = Problem2(input);
    assert(ret2 == 15343601);
    printf("ret1 = %d\n", ret1);
    printf("ret2 = %d\n", ret2);
    return 0;
}

File: 2015/python/day06.py
#!/usr/bin/env python3
from enum import Enum
import dataclasses


class Action(Enum):
    ON = 1
    OFF = 2
    TOGGLE = 3


@dataclasses.dataclass
class Data:
    action: Action
    start: tuple[int, int]
    end: tuple[int, int]


def parse(s: str) -> Data:
    parts = s.split(None)
    if parts[0] == "turn":
        action = Action.ON if parts[1] == "on" else Action.OFF
        starts = parts[2].split(",")
        ends = parts[4].split(",")
        return Data(
            action, (int(starts[0]), int(starts[1])), (int(ends[0]), int(ends[1]))
        )
    else:
        starts = parts[1].split(",")
        ends = parts[3].split(",")
        return Data(
            Action.TOGGLE,
            (int(starts[0]), int(starts[1])),
            (int(ends[0]), int(ends[1])),
        )


def problem1(actions: list[Data]) -> int:
    matrix = [[False] * 1000 for _ in range(0, 1000)]
    for d in actions:
        match d.action:
            case Action.ON:
                for x in range(d.start[0], d.end[0] + 1):
                    for y in range(d.start[1], d.end[1] + 1):
                        matrix[x][y] = True
            case Action.OFF:
                for x in range(d.start[0], d.end[0] + 1):
                    for y in range(d.start[1], d.end[1] + 1):
                        matrix[x][y] = False
            case Action.TOGGLE:
                for x in range(d.start[0], d.end[0] + 1):
                    for y in range(d.start[1], d.end[1] + 1):
                        matrix[x][y] = not matrix[x][y]

    return sum([len([x for x in row if x]) for row in matrix])


def problem2(actions: list[Data]) -> int:
    matrix = [[0] * 1000 for _ in range(0, 1000)]
    for d in actions:
        match d.action:
            case Action.ON:
                for x in range(d.start[0], d.end[0] + 1):
                    for y in range(d.start[1], d.end[1] + 1):
                        matrix[x][y] += 1
            case Action.OFF:
                for x in range(d.start[0], d.end[0] + 1):
                    for y in range(d.start[1], d.end[1] + 1):
                        matrix[x][y] = max(0, matrix[x][y] - 1)
            case Action.TOGGLE:
                for x in range(d.start[0], d.end[0] + 1):
                    for y in range(d.start[1], d.end[1] + 1):
                        matrix[x][y] += 2

    return sum([sum(row) for row in matrix])


with open("../input/day06.txt") as f:
    input = [parse(line) for line in f.readlines()]
    ret1 = problem1(input)
    assert ret1 == 400410

    ret2 = problem2(input)
    assert ret2 == 15343601

    print(f"ret1={ret1}")
    print(f"ret2={ret2}")
